lib/web/apiserver.go

```diff
@@ -2142,8 +2142,56 @@ func (h *Handler) siteNodeConnect(
 		return nil, trace.Wrap(err)
 	}
 
+	// Fetching nodes is quite slow when many nodes exist so fetch it once and
+	// pass the value to the methods that require it.
+	clt, err := ctx.GetUserClient(site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	namespace := apidefaults.Namespace
+	if req.Namespace != "" {
+		namespace = req.Namespace
+	}
+	servers, err := clt.GetNodes(r.Context(), namespace)
```

Comment:
This is where you said in the happy path we could dial by UUID and remove this `GetNodes` as well, right? If so, let's add a TODO for later.